osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/datadecoder/BasicDlmsDataDecoder.java

```diff
@@ -0,0 +1,166 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service(value = "basicDlmsDataDecoder")
+public class BasicDlmsDataDecoder {
+
+  // This service decodes attributes that have a basic DLMS data type.
+
+  private static final List<DataObject.Type> simpleNumbers =
+      List.of(
+          DataObject.Type.DOUBLE_LONG,
+          DataObject.Type.DOUBLE_LONG_UNSIGNED,
+          DataObject.Type.INTEGER,
+          DataObject.Type.UNSIGNED,
+          DataObject.Type.LONG64,
+          DataObject.Type.LONG64_UNSIGNED,
+          DataObject.Type.LONG_INTEGER,
+          DataObject.Type.LONG_UNSIGNED);
+
+  private final DlmsHelper dlmsHelper;
+
+  @Autowired
+  public BasicDlmsDataDecoder(final DlmsHelper dlmsHelper) {
+    this.dlmsHelper = dlmsHelper;
+  }
+
+  public String decodeAttributeValue(final DataObject attributeData)
+      throws ProtocolAdapterException {
+
+    if (simpleNumbers.contains(attributeData.getType())) {
+      return this.decodeNumber(attributeData);
+    }
+
+    if (attributeData.getType() == Type.BOOLEAN) {
+      return this.decodeBoolean(attributeData);
+    }
+
+    if (attributeData.getType() == Type.OCTET_STRING) {
+      return this.decodeOctetString(attributeData);
+    }
+
+    if (attributeData.getType() == Type.VISIBLE_STRING) {
+      return this.decodeVisibleString(attributeData);
+    }
+
+    return null;
+  }
+
+  private String decodeNumber(final DataObject attributeData) throws ProtocolAdapterException {
+    final Long value = this.dlmsHelper.readLong(attributeData, "read number");
+
+    if (value == null) {
+      return "null";
+    } else {
+      return value.toString();
+    }
+  }
+
+  private String decodeBoolean(final DataObject attributeData) {
+    return attributeData.getValue().toString();
+  }
+
+  private String decodeOctetString(final DataObject attributeData) {
+
+    try {
+      return this.decodeDateTime(attributeData);
+    } catch (final Exception ignored) {
+      // Decoding failed, so byte array is not a date time
+    }
+
+    final byte[] byteArray = attributeData.getValue();
+    return new String(byteArray);
+  }
+
+  public String decodeDateTime(final DataObject attributeData) throws ProtocolAdapterException {
+    return this.cosemDateTimeToString(
+        this.dlmsHelper.readDateTime(attributeData, "Try to read time from bytes"));
+  }
+
+  public String decodeDate(final DataObject attributeData) {
+    if (attributeData.getType() == Type.OCTET_STRING) {
+
+      final ByteBuffer bb = ByteBuffer.wrap(attributeData.getValue());
+
+      final int year = bb.getShort() & 0xFFFF;
+      final int month = bb.get() & 0xFF;
+      final int dayOfMonth = bb.get() & 0xFF;
+      final int dayOfWeek = bb.get() & 0xFF;
+
+      return this.getDayOfWeek(dayOfWeek)
+          + ", "
+          + ((year == 65535) ? "Year not specified" : year)
```

Comment:
2 ^ 16

Reply:
Changed to 0xFFFF, then it's the same as in the DLMS specification: `0xFFFF = not specified`